nix/mobile/android/maven-and-npm-deps/default.nix

```diff
@@ -0,0 +1,207 @@
+          substituteInPlace android/gradle.properties \
+            --replace 'org.gradle.jvmargs=-Xmx8704M' 'org.gradle.jvmargs=-Xmx8704M
+          org.gradle.daemon=false
+          org.gradle.caching=false'
+
+          set $prevSet
+
+          runHook postPatch
+        '';
+        buildPhase =
+          androidEnvShellHook +
+          status-go.shellHook + ''
+          export HOME=$NIX_BUILD_TOP
+          export REACT_NATIVE_DEPENDENCIES="${reactNativeDepsDir}"
+          export STATUS_REACT_HOME="${projectDir}"
+
+          pushd ${projectDir}/android
+          # NOTE: This generates the react-native-android binaries under node_modules/react-native/android
+          LD_LIBRARY_PATH=$LD_LIBRARY_PATH:${lib.makeLibraryPath [ zlib ]} \
+            gradle -Dmaven.repo.local=${mavenRepoDir} --offline -S react-native-android:installArchives || exit
+          popd > /dev/null
+        '';
+        doCheck = true;
+        checkPhase = ''
+          runHook preCheck
+          test -d ${projectDir}/node_modules/react-native/ReactAndroid/build/intermediates/javac/release/compileReleaseJavaWithJavac/classes/com/facebook || \
+            exit 1
+          test -d ${projectDir}/node_modules/react-native/ReactAndroid/build/react-ndk/exported || \
+            exit 2
+          runHook postCheck
+        '';
+        installPhase = ''
+          rm -rf $out
+          mkdir -p $out/project $out/.m2/repository
+
+          # TODO: maybe node_modules/react-native/ReactAndroid/build/{tmp,generated} can be discarded?
+          cp -R ${mavenRepoDir} $out/.m2/
+          cp -R ${projectDir}/android/ ${projectDir}/node_modules/ $out/project
+        '';
+        fixupPhase = ''
+          # Patch prepareJSC so that it doesn't subsequently try to build NDK libs
+          substituteInPlace $out/project/node_modules/react-native/ReactAndroid/build.gradle \
+            --replace 'packageReactNdkLibs(dependsOn: buildReactNdkLib, ' 'packageReactNdkLibs('
+        '';
+
+        # The ELF types are incompatible with the host platform, so let's not even try
+        # TODO: Use Android NDK to strip binaries manually
+        dontPatchELF = true;
+        dontStripHost = true;
+
+        # Take whole sources into consideration when calculating sha
+        outputHashMode = "recursive";
+        outputHashAlgo = "sha256";
+      };
+
+in {
+  buildInputs = deps;
+
+  shellHook = ''
+    # This avoids RN trying to download dependencies. Maybe we need to wrap this in a special RN environment derivation
+    ln -sf $STATUS_REACT_HOME/mobile_files/package.json.orig $STATUS_REACT_HOME/package.json
+    ln -sf $STATUS_REACT_HOME/mobile_files/metro.config.js $STATUS_REACT_HOME/metro.config.js
+    ln -sf $STATUS_REACT_HOME/mobile_files/yarn.lock $STATUS_REACT_HOME/yarn.lock
+
+    export STATUSREACT_NIX_MAVEN_REPO="${deps}/.m2/repository"
+  '';
+}
```

Comment:
I personally think the risk of this breaking is relatively low, since the way we're interacting with platforms involved (Maven, Gradle) doesn't change that often. The bigger risk is that react-native/ReactAndroid changes its gradle files (I don't expect radical changes in this type of code either), but that'll be part of any future RN upgrade process.

Reply:
I'm not sure what you mean by "for the benefit that we get". Do you mean reproducible builds as a whole?